drivers/hpet.c

```diff
@@ -0,0 +1,96 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <drivers/hpet.h>
+#include <ioapic.h>
+
+bool init_hpet(uint8_t dst_cpus) {
+    printk("Initializing HPET\n");
+    acpi_hpet_t *hpet = (acpi_hpet_t *) acpi_find_table(HPET_SIGNATURE);
+    if (hpet) {
+        mfn_t hpet_base_mfn = paddr_to_mfn(hpet->address.address);
+        vmap((uint64_t *) hpet->address.address, hpet_base_mfn, PAGE_ORDER_4K, L1_PROT);
+        volatile acpi_hpet_timer_t *config =
+            (acpi_hpet_timer_t *) (hpet->address.address +
+                                   ACPI_HPET_OFFSET_TIMER_0_CONFIG_CAP_REG);
+        volatile acpi_hpet_general_t *general =
+            (acpi_hpet_general_t *) (hpet->address.address +
+                                     ACPI_HPET_OFFSET_GENERAL_CAP_REG);
+        volatile uint64_t *main_counter =
+            (uint64_t *) (hpet->address.address +
+                          ACPI_HPET_OFFSET_GENERAL_MAIN_COUNTER_REG);
+
+        general->config.enabled = 0;
+
+        if (config->cfg_int_type != ACPI_HPET_CONFIG_INT_TYPE_TRIGGER_MODE) {
+            return false;
+        }
+
+        if (!config->cap_int_periodic) {
+            return false;
+        }
+
+        if (general->cap.num_timers_cap == 0) {
+            return false;
+        }
+
+        int i;
+        acpi_hpet_timer_t local;
+        /* Disable all timers */
+        for (i = 0; i < general->cap.num_timers_cap; ++i) {
+            local = *(config + i);
+            local.cfg_int_enabled = 0;
+            *(config + i) = local;
+        }
+
+        *main_counter = 0;
+        local = *config;
+
+        /* 1 fs = 10^15s */
+        uint64_t freq_hz = 1000000000000000 / general->cap.counter_clock_period;
+        uint64_t ticks = freq_hz / 1000; /* Interrupt every 1ms */
+
+        local.cfg_int_enabled = 1;
+        local.cfg_int_route = 0;
+        local.cfg_fsb_enable = 0;
+        local.cfg_type = 1;
+        local.cfg_value_set = 1;
+        local.comparator = ticks;
+
+        *config = local;
+
+        general->config.leg_repl_cfg = 0; /* Disable legacy route */
+        general->config.enabled = 1;
+
+        if (general->config.enabled && config->cfg_int_enabled) {
+            configure_isa_irq(PIT_IRQ, PIT_IRQ0_OFFSET, IOAPIC_DEST_MODE_PHYSICAL,
```

Comment:
maybe define dedicated HPET IRQ and OFFSET macros (they may overlap with PIT if this is wanted).